### PR TITLE
SignalFx stats exporter

### PR DESCRIPTION
**Pull request by mpetazzoni**
A stats exporter to report stats as SignalFx metric timeseries.

Note that SignalFx doesn't have native histogram support yet, so those views can't be reported to SignalFx and are ignored for now.

### CODE REVIEW

**Comment by adriancole** [comment]
Ps I keep hearing good things about your product!

**Comment by mpetazzoni** [comment]
@adriancole Thanks! I'm very proud of the technology and product that we've built and I'm super excited about all the new things we're working on!

I've signed a company-wide CLA for SignalFx; waiting to here back from `cla-submissions`.

**Comment by mpetazzoni** [comment]
So I guess the CI build isn't happy about the `opencensusVersion` in `opencensus-examples`. But since `opencensus-examples` is now a distinct Gradle build, I'm not sure how to handle the chicken and egg problem I mentioned in the PR. Let me know how you'd want this handled and I'll make the required changes.

**Comment by bogdandrutu** [comment]
@mpetazzoni Try to resend the PR for CLA. For the errors I see some:
java/exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java:96: error: [ReturnMissingNullable] Returning null literal
    return null;
(see http://errorprone.info/bugpattern/ReturnMissingNullable)
  Did you mean '@Nullable'?

**Comment by mpetazzoni** [comment]
Interesting, I ran `./gradlew clean assemble check` like the `CONTRIBUTING.md` document suggested, and it didn't complain about that.

File: exporters/stats/signalfx/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  </dependency>
+  <dependency>
+    <groupId>io.opencensus</groupId>
+    <artifactId>opencensus-impl</artifactId>
+    <version>0.10.0</version>
+  </dependency>
+  <dependency>
+    <groupId>io.opencensus</groupId>
+    <artifactId>opencensus-exporter-stats-signalfx</artifactId>
+    <version>0.10.0</version>
```
Comment (bogdandrutu) [inline exporters/stats/signalfx/README.md]: the correct version is 0.10.1

File: examples/src/main/java/io/opencensus/examples/stats/SignalFxStatsExample.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * Example application that reports its own heap usage as stats to SignalFx.
+ *
+ * <p>Measures every second the JVM's heap usage, maximum heap and an iteration count and reports
+ * those as 'jvm.heap.used', 'jvm.heap.max' and 'iterations' metrics. In this simple example, we use
+ * the measure name as the metric name because we only have one view on each of those measures.
+ *
+ * <p>This program runs forever until ^C-ed.
+ */
+public class SignalFxStatsExample {
```
Comment (bogdandrutu) [inline examples/src/main/java/io/opencensus/examples/stats/SignalFxStatsExample.java]: Can we move this into a different commit (and all the changes in examples)?

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+              .addAllDimensions(createDimensions(keys, entry.getKey()))
+              .setValue(createDatum(entry.getValue()))
+              .build());
+    }
+    return datapoints;
+  }
+
+  @VisibleForTesting
+  @Nullable
+  static MetricType getMetricTypeForAggregation(Aggregation aggregation, AggregationWindow window) {
```
Comment (bogdandrutu) [inline exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: I think we have the following windows:
* AggregationWindow.Interval - supports things like last minute, last hour. this is useful for in-process debugging usually, not good to export.
* AggregationWindow.Cumulative - this is the total since the registration of the view.
* AggregationWindow.Delta - we will support this in the future and will be the total since the last export.

In case of a AggregationWindow.Cumulative I think (unless you support sending cumulative data) all our aggregation can be sent as GAUGEs.

What I think you have in CUMULATIVE_COUNTER is actually what we call AggregationWindow.Delta (correct me if I am wrong).

I am not able to fully understand your COUNTER metric. Do you send deltas or total?

File: examples/src/main/java/io/opencensus/examples/stats/SignalFxStatsExample.java (hunk 2 above)
Comment (mpetazzoni) [reply examples/src/main/java/io/opencensus/examples/stats/SignalFxStatsExample.java]: Yup, I think I don't have a choice anyway because it's not going to build otherwise. I'll send a different PR later with the example once this is merged. Note that it doesn't seem possible to implement anything in `opencensus-examples` that depends on an unreleased version of `opencensus-java`. It might also make sense to break out the Java examples into their own repo to avoid the confusion between the builds.

File: exporters/stats/signalfx/README.md (hunk 1 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/README.md]: Right, missed that in the rebase. Will fix.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java (hunk 3 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: Cumulative counters send monotonically increasing values, for which the absolute value doesn't matter; similar to SNMP counters. We can then calculate a time-adjusted delta to get a rate per second. Cumulative counters are allowed to rollover (for example on service restart, or overflow).

For "pure" counters, only deltas get sent. This type of counter is usually less tolerant of loss: if a datapoint doesn't make it, you loose that increment. With cumulative counters, the next value will set you back straight. For this reason, "pure" counters are usually only used for distributed counting (multiple emitters contributing increments to the same counter timeseries).

Sounds like I don't want to deal with Interval, but I should be supporting Delta. I'll fix that.

Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: @bogdandrutu It doesn't look like `AggregationWindow.Delta` exists yet.

Comment (bogdandrutu) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: @mpetazzoni as I said AggregationWindow.Delta will be supported in the future version. I will let you know when available.

Comment (bogdandrutu) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: @mpetazzoni I am waiting for you to fix the types based on the discussion.

**Comment by mpetazzoni** [comment]
I'll continue to work on the test coverage. I'm also still waiting to hear back from `cla-submissions` on the company-wide CLA for SignalFx.

**Comment by mpetazzoni** [comment]
I signed it!

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+  @VisibleForTesting
+  static Iterable<Dimension> createDimensions(List<TagKey> keys, List<TagValue> values) {
+    Preconditions.checkArgument(
+        keys.size() == values.size(), "TagKeys and TagValues don't have the same size.");
+    List<Dimension> dimensions = new ArrayList<>(keys.size());
+    for (ListIterator<TagKey> it = keys.listIterator(); it.hasNext(); ) {
+      TagKey key = it.next();
+      TagValue value = values.get(it.previousIndex());
+      if (value == null || Strings.isNullOrEmpty(value.asString())) {
```
Comment (songy23) [inline exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: Just curious, SignalFx doesn't allow empty string as Dimension value? (In OpenCensus using "" as TagValue is OK.)

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+   * Creates a SignalFx Stats exporter that gets its authentication token and the SignalFx ingest
+   * URL from the given Java properties and uses the default reporting interval.
+   *
+   * <p>Only one SignalFx Stats exporter can be created.
+   *
+   * @param properties a {@code Properties} object containing the {@link #SIGNALFX_TOKEN_PROPERTY}
+   *     and {@link #SIGNALFX_HOST_PROPERTY}.
+   * @throws IllegalStateException if a SignalFx exporter already exists.
+   */
+  public static void create(Properties properties) {
```
Comment (songy23) [inline exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: nit: checkNotNull(properties, "properties");

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+   * Creates a SignalFx Stats exporter with an explicit authentication token and reporting interval.
+   *
+   * <p>Only one SignalFx Stats exporter can be created.
+   *
+   * @param host the SignalFx ingest URL.
+   * @param token a SignalFx ingest token.
+   * @param interval the interval at which stats are reported to SignalFx.
+   * @throws IllegalStateException if a SignalFx exporter already exists.
+   */
+  public static void create(String host, String token, Duration interval) {
```
Comment (songy23) [inline exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: nit: checkNotNull(interval, "interval");

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThread.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    setName(getClass().getSimpleName());
+    logger.log(Level.FINE, "Initialized SignalFx exporter to {0}.", url);
+  }
+
+  @VisibleForTesting
+  void export() {
+    try (Session session = sender.createSession()) {
+      for (View view : views.getAllExportedViews()) {
+        ViewData data = views.getView(view.getName());
+        for (DataPoint datapoint : SignalFxSessionAdaptor.adapt(data)) {
```
Comment (songy23) [inline exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThread.java]: nit: could also add all datapoints generated from all view data to one list, and export them all at once?

File: exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptorTest.java [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+  @Test
+  public void createDatumFromDistributionThrows() {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Distribution aggregations are not supported");
+    SignalFxSessionAdaptor.createDatum(
+        DistributionData.create(5, 2, 0, 10, 40, ImmutableList.of(1L)));
+  }
+
+  @Test
+  public void adaptViewIntoDatapoints() {
```
Comment (songy23) [inline exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptorTest.java]: suggestion: consider adding another test where tag values have null or empty string.

File: exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterTest.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  private static final String TEST_ENDPOINT = "https://example.com";
+  private static final Duration ONE_SECOND = Duration.create(1, 0);
+
+  @Rule public final ExpectedException thrown = ExpectedException.none();
+
+  @After
+  public void tearDown() {
+    System.getProperties().remove(SignalFxStatsExporter.SIGNALFX_HOST_PROPERTY);
+    System.getProperties().remove(SignalFxStatsExporter.SIGNALFX_TOKEN_PROPERTY);
+    SignalFxStatsExporter.unsafeResetExporter();
```
Comment (songy23) [inline exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterTest.java]: I think unsafeResetExporter() needs to be called after each test cases that creates an exporter.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java (hunk 4 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: We don't. If there's no value, the dimension key/value shouldn't be there.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxMetricsSenderFactory.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+import com.signalfx.metrics.auth.StaticAuthToken;
+import com.signalfx.metrics.connection.HttpDataPointProtobufReceiverFactory;
+import com.signalfx.metrics.connection.HttpEventProtobufReceiverFactory;
+import com.signalfx.metrics.errorhandler.OnSendErrorHandler;
+import com.signalfx.metrics.flush.AggregateMetricSender;
+import java.net.URL;
+import java.util.Collections;
+
+/** Interface for creators of {@link AggregateMetricSender}. */
+public interface SignalFxMetricsSenderFactory {
```
[the remark this answers is not in the document]
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxMetricsSenderFactory.java]: Yup, definitely could. IDE default. Fixed.

File: exporters/stats/signalfx/README.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+```xml
+<dependencies>
+  <dependency>
+    <groupId>io.opencensus</groupId>
+    <artifactId>opencensus-api</artifactId>
+    <version>0.10.1</version>
+  </dependency>
+  <dependency>
+    <groupId>io.opencensus</groupId>
+    <artifactId>opencensus-impl</artifactId>
```
[the remark this answers is not in the document]
Comment (mpetazzoni) [reply exporters/stats/signalfx/README.md]: Good catch, fixed.

File: exporters/stats/signalfx/README.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    <version>0.10.1</version>
+  </dependency>
+  <dependency>
+    <groupId>io.opencensus</groupId>
+    <artifactId>opencensus-exporter-stats-signalfx</artifactId>
+    <version>0.10.1</version>
+    <scope>runtime</scope>
```
[the remark this answers is not in the document]
Comment (mpetazzoni) [reply exporters/stats/signalfx/README.md]: Ditto

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java (hunk 5 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: Added

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java (hunk 6 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: Added to constructor, checking parameters there.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThread.java (hunk 7 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThread.java]: The export already happens only once: when the session is closed at the end of the try-with-resource block.

File: exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterTest.java (hunk 9 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterTest.java]: Not sure I see your point. That's exactly what this `@After` method does. For tests where the exporter isn't created (expected failure), calling `unsafeResetExporter()` will simply be a no-op.

File: exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptorTest.java (hunk 8 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptorTest.java]: I have a distinct unit test for that already but I added another test all the way to `Datapoint` creation.

File: exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterTest.java (hunk 9 above)
Comment (songy23) [reply exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterTest.java]: Ahh that's right, I mixed up with `@After`, `@AfterEach` and `@AfterClass`.

**Comment by mpetazzoni** [comment]
Thanks for the quick turnaround!

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
```
Comment (sebright) [inline exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: Suggestion: We just switched to using a configuration class with a builder in the StackDriver exporter (#896).  That design might work well here.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+ *   SignalFxStatsExporter.create();
+ *   ... // Do work.
+ * }
+ * </code></pre>
+ */
+public final class SignalFxStatsExporter {
+
+  public static final String SIGNALFX_TOKEN_PROPERTY = "signalfx.auth.token";
+  public static final String SIGNALFX_HOST_PROPERTY = "signalfx.ingest.url";
+  public static final String DEFAULT_SIGNALFX_ENDPOINT = "https://ingest.signalfx.com";
```
Comment (sebright) [inline exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: Should these constants have Javadocs?

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThread.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+    setDaemon(true);
+    setName(getClass().getSimpleName());
+    logger.log(Level.FINE, "Initialized SignalFx exporter to {0}.", url);
+  }
+
+  @VisibleForTesting
+  void export() {
+    try (Session session = sender.createSession()) {
+      for (View view : views.getAllExportedViews()) {
```
Comment (sebright) [inline exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThread.java]: Should this method skip exporting the Distribution views for now?

**Comment by sebright** [comment]
@mpetazzoni - @dinooliva, @bogdandrutu, and I probably won't have time to finish reviewing your PR until January.

File: exporters/stats/signalfx/README.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  </dependency>
+</dependencies>
+```
+
+For Gradle add to your dependencies:
+
+```
+compile 'io.opencensus:opencensus-api:0.10.1'
+compile 'io.opencensus:opencensus-impl:0.10.1'
+runtime 'io.opencensus:opencensus-exporter-stats-signalfx:0.10.1'
```
[the remark this answers is not in the document]
Comment (mpetazzoni) [reply exporters/stats/signalfx/README.md]: Good catch; fixing.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java (hunk 14 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: Will add.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java (hunk 13 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: I'll look into this. The configuration here seems a lot simpler than for the StackDriver exporter, there really only is the token to specify, and the simplest way to do that is via Java system properties which can be wired through by the orchestration/deployment system (passing in a `-Dsignalfx.auth.token=`).

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThread.java (hunk 15 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThread.java]: The adaptor does that already. I didn't want to have logic for this in the worker thread to keep this class as simple as it needs to be, with the adaptor having the understanding of what views contain and how to deal with them.

File: exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThreadTest.java [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+    Mockito.when(session.setDatapoint(Mockito.any(DataPoint.class))).thenReturn(session);
+    Mockito.when(viewManager.getAllExportedViews()).thenReturn(ImmutableSet.<View>of());
+
+    SignalFxStatsExporterWorkerThread thread =
+        new SignalFxStatsExporterWorkerThread(
+            factory, endpoint, TEST_TOKEN, ONE_SECOND, viewManager);
+    thread.start();
+    thread.interrupt();
+    thread.join();
+    assertFalse("Worker thread should have stopped", thread.isAlive());
```
Comment (sebright) [inline exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThreadTest.java]: It might be better to add a timeout to this test or the call to `join`, since `join` should only return normally if `isAlive` returns false.

File: exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThreadTest.java [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+    Mockito.when(view.getAggregation()).thenReturn(Aggregation.Mean.create());
+    Mockito.when(view.getWindow()).thenReturn(AggregationWindow.Cumulative.create());
+    Mockito.when(view.getColumns()).thenReturn(ImmutableList.of(TagKey.create("animal")));
+
+    ViewData viewData = Mockito.mock(ViewData.class);
+    Mockito.when(viewData.getView()).thenReturn(view);
+    Mockito.when(viewData.getAggregationMap())
+        .thenReturn(
+            ImmutableMap.<List<TagValue>, AggregationData>of(
+                ImmutableList.of(TagValue.create("cat")), SumDataDouble.create(3.14d)));
```
Comment (sebright) [inline exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThreadTest.java]: It looks like the Aggregation and AggregationData are inconsistent.  One is "mean" and the other is "sum".

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+   * Converts the given view data into datapoints that can be sent to SignalFx.
+   *
+   * <p>The view name is used as the metric name, and the aggregation type and aggregation window
+   * type determine the metric type.
+   *
+   * @param data The {@link ViewData} containing the aggregation data of each combination of tag
+   *     values.
+   * @return A list of datapoints for the corresponding metric timeseries of this view's metric.
+   */
+  static List<DataPoint> adapt(@Nonnull ViewData data) {
```
[the remark this answers is not in the document]
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: Agreed. Will fix.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+   * @return a {@code Builder}.
+   */
+  public static Builder builder() {
+    try {
+      return new AutoValue_SignalFxStatsConfiguration.Builder()
+          .setIngestEndpoint(new URI(DEFAULT_SIGNALFX_ENDPOINT))
+          .setExportInterval(DEFAULT_EXPORT_INTERVAL);
+    } catch (URISyntaxException e) {
+      // This shouldn't happen if DEFAULT_SIGNALFX_ENDPOINT was typed in correctly.
+      throw new IllegalArgumentException(e);
```
[the remark this answers is not in the document]
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsConfiguration.java]: Indeed. I was under the (incorrect) impression that I couldn't make a field final if it was initialized in a static block. But since that works, I'll fix that.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+  @GuardedBy("monitor")
+  @Nullable
+  private static SignalFxStatsExporter exporter = null;
+
+  private SignalFxStatsExporter(SignalFxStatsConfiguration configuration, ViewManager viewManager) {
+    Preconditions.checkNotNull(configuration, "SignalFx stats exporter configuration");
+    this.configuration = configuration;
+
+    String token = configuration.getToken();
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(token), "Invalid SignalFx token");
```
[the remark this answers is not in the document]
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: Indeed; didn't realize this was possible with AutoValue, but https://github.com/google/auto/blob/master/value/userguide/builders-howto.md#validate has the solution.

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+
+    Duration interval = configuration.getExportInterval();
+    Preconditions.checkArgument(interval.compareTo(ZERO) > 0, "Interval duration must be positive");
```
[the remark this answers is not in the document]
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporter.java]: Ack

File: exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThreadTest.java (hunk 18 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThreadTest.java]: Fixed

File: exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThreadTest.java (hunk 17 above)
Comment (mpetazzoni) [reply exporters/stats/signalfx/src/test/java/io/opencensus/exporter/stats/signalfx/SignalFxStatsExporterWorkerThreadTest.java]: Good point. Done.

**Comment by mpetazzoni** [comment]
@bogdandrutu, @dinooliva, any other comments on this? If not, it's ready to be squash & merged whenever you want! Thanks!

**Comment by mtwo** [comment]
FYI, Bogdan's still on winter vacation but will be back next week

File: exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+    return datapoints;
+  }
+
+  @VisibleForTesting
+  @javax.annotation.Nullable
+  static MetricType getMetricTypeForAggregation(Aggregation aggregation, AggregationWindow window) {
+    if (aggregation instanceof Aggregation.Mean) {
+      return MetricType.GAUGE;
+    } else if (aggregation instanceof Aggregation.Count || aggregation instanceof Aggregation.Sum) {
+      if (window instanceof AggregationWindow.Interval) {
```
Comment (bogdandrutu) [inline exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: Simply ignore for the moment the interval aggregation window, it is better.

Comment (mpetazzoni) [reply exporters/stats/signalfx/src/main/java/io/opencensus/exporter/stats/signalfx/SignalFxSessionAdaptor.java]: Done, we'll leave this out for now. Easy enough to put back in when interval aggregation window is ready to be used.

**Comment by sebright** [comment]
Thanks! I'll squash and merge the PR when the build passes.

**Comment by mpetazzoni** [comment]
Yup, sorry about that. Forgot a file in my commit.

**Comment by songy23** [comment]
Thanks! I merged it.
